### DISCUSSION TITLE
Best way to map out DHCP ranges

### DISCUSSION THREAD

**Discussion by robertmain** [netbox-community/netbox#13441] [Q&A] [answered]
I'd like to start documenting my DHCP ranges in Netbox, but I'm not sure of the best (or correct) way to do that. I _could_ do it by bulk creating IPs, but of course those are then going to be counted against the utilization of whatever prefix I create them under. I _could_ also do it as an IP Range, but that then shows 0% utilization for the range which isn't necessarily correct either (a given DHCP range, especially for guests could be be anywhere between 0% and 100% utilized at a given moment). I did notice something along those lines was alluded to [here](https://github.com/netbox-community/netbox/discussions/7560)

I'd love to know if there's an accepted way to do this in Netbox (there might be, and I just missed it in the docs).

**Comment by mtinberg** [comment, marked as answer]
I don't think you need to get hung up on the real-time DHCP utilization figures which would be known only to the DHCP server software, Netbox is good for storing data about the intended configuration of the network, so the fact that an IP Range is allocated for dynamic DHCP is useful for configuring your DHCP service, and knowing those IPs have a purpose (they're assigned to DHCP), but real-time data you can get from other monitoring tools does not need to be in Netbox as its not needed for templating and pushing config.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Robert Main ***@***.***>
Sent: Thursday, August 10, 2023 4:43 PM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Best way to map out DHCP ranges (Discussion #13441)


I'd like to start documenting my DHCP ranges in Netbox, but I'm not sure of the best (or correct) way to do that. I could do it by bulk creating IPs, but of course those are then going to be counted against the utilization of whatever prefix I create them under. I could also do it as an IP Range, but that then shows 0% utilization for the range which isn't necessarily correct either (a given DHCP range, especially for guests could be be anywhere between 0% and 100% utilized at a given moment).

I'd love to know if there's an accepted way to do this in Netbox (there might be, and I just missed it in the docs).

**Comment by robertmain** [comment]
Yeah, I'm not too hung up on real time data per-se. I just brought up the utilization because it's not entirely accurate to call the range either 0% utilized or 100% utilized. Though at the same time, you could argue that it's 100% _allocated_ (a.k.a: utilized). Should I use IP ranges for this and then just mark it allocated?

**Reply by zndrr** [reply]
This is how I have done it, so would recommend that approach. For static mappings I enter those too, but that's out of the dynamic scope anyway.

**Reply by candlerb** [reply]
I agree. The utilization % then shows the amount of space allocated in the prefix, including the DHCP pool at worst-case if it is fully utilized.

This figure gives you an overall idea how much leeway you have, either to add individual hosts with static IPs, or to grow the DHCP pool if it becomes necessary.

The actual minimum and maximum utilization of the DHCP pool itself (and therefore whether the pool is correctly sized) is not really something that Netbox itself can or should monitor.

**Reply by robertmain** [reply]
Looking at this again, it looks like I can select "Role=DHCP" for individual IPs, but not IP ranges. Would it make sense for me to switch to using individual IP addresses in that instance? I say that simply because without that, it makes it a bit harder to use this information when generating configuration templates. 

There's also a part of me that feels like I'm being nudged toward using individual IPs rather than a range since individual IPs have the option built in to select "DHCP" whereas ranges do not (unless you add a custom role for that which is what I have right now).

**Reply by candlerb** [reply]
Individual IPs was the way you had to do it in Netbox before Ranges were added.  For a pool of 100 IPv4 addresses, it was workable. For a pool of 1,000,000 IPv6 addresses, not so much :-)

That's why the DHCP IPAddress role is still there - and you can continue to work this way if you prefer.  You need to take a continuous range of adjacent IP addresses and convert them into a start and end, but I think there are some helper functions for this.

Range roles don't default to the same set as IP Roles because a lot of them wouldn't make sense for a Range (e.g. VRRP, Anycast). I don't find the current defaults of "Backup", "Development", "Management", "Production" useful, but you can customize them.

**Comment by bitcollector1** [comment]
I'm using custom fields to track DHCP Pools 

Also using KEA to find DHCP utilization....I'm hoping to work on a plugin before too long. 

```
from pykeadhcp import Kea

server = Kea(host="https://kea.vip.hwe.com", port=443, use_basic_auth=True, verify = "/etc/ssl/certs",
             username="user", password="admin")

subnet_id = [73]

for i in subnet_id:

    total_dhcp = list(server.dhcp4.statistic_get(f"subnet[{i}].total-addresses").arguments.values())
    total_dhcp = total_dhcp[0][0][0]
    total_dhcp

    assigned_dhcp = list(server.dhcp4.statistic_get(f"subnet[{i}].assigned-addresses").arguments.values())
    assigned_dhcp = assigned_dhcp[0][0][0]
    assigned_dhcp

    dhcp_utilization = (assigned_dhcp / total_dhcp) * 100 
    print(f"{int(dhcp_utilization)}% DHCP Utilization")
```
```
24% DHCP Utilization
```

This is actually an incorrect figure because I believe it does not take into account "Host Reservations" in that subnet, so It still needs to be tweaked to account for that scenario.

**Comment by bitcollector1** [comment]
This plugin is awesome, but I need to to do more, so I think I will use it as an example as I work through my own plugin 

https://github.com/devon-mar/netbox-kea

<img width="1326" alt="Screenshot 2023-08-17 at 08 23 45" src="https://github.com/netbox-community/netbox/assets/50723251/348dc6cf-c055-4b49-9531-a197fbdc1966">

**Comment by bitcollector1** [comment]
Just did some more reading and implemented this IP Range concept, thanks for the food for thought :)

**Comment by bitcollector1** [comment]
Is the "Utilization" supposed to be available for these ranges via API?  I'm not seeing it 

![Screenshot 2023-08-22 at 17 41 29](https://github.com/netbox-community/netbox/assets/50723251/6b344554-e811-46eb-a13a-953831ac13e8)

```
{
    "id": 28,
    "url": "https://netdevbox.hwe.ebay.com/api/ipam/ip-ranges/28/",
    "display": "10.254.151.51-62/28",
    "family": {
        "value": 4,
        "label": "IPv4"
    },
    "start_address": "10.254.151.51/28",
    "end_address": "10.254.151.62/28",
    "size": 12,
    "vrf": null,
    "tenant": {
        "id": 39,
        "url": "https://netdevbox.hwe.ebay.com/api/tenancy/tenants/39/",
        "display": "infra",
        "name": "infra",
        "slug": "infra"
    },
    "status": {
        "value": "reserved",
        "label": "Reserved"
    },
    "role": null,
    "description": "1105A-105-MA (DHCP Pool)",
    "comments": "",
    "tags": [],
    "custom_fields": {},
    "created": "2023-08-22T09:03:02.058620-07:00",
    "last_updated": "2023-08-22T14:12:09.168663-07:00",
    "mark_utilized": false
}

```

**Comment by bitcollector1** [comment]
I would also think 100% would be Dark Red, I actually glanced right over the ranges that were in this state  :)   

<img width="252" alt="Screenshot 2023-08-22 at 17 45 58" src="https://github.com/netbox-community/netbox/assets/50723251/4703d68e-6375-498a-a1cf-b74eceb91788">
